### PR TITLE
Feature/podspec v2 cm secrets rbac

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

a few new features and enhancements for `podspecV2`:
- Added `secrets`, `configmap` support
- Changed `RBAC` logic a bit for the new spec;

## QA steps

- deploy some a charm using podspec v2(has `RBAC`, `secrets` and `configmap`);

## Documentation changes

Yes

## Bug reference

None

### CODE REVIEW

File: caas/kubernetes/provider/k8s_test.go [hunk 1]
```diff
@@ -1969,44 +1956,43 @@ func (s *K8sBrokerSuite) TestEnsureServiceWithServiceAccountNewClusterRoleUpdate
...
-		s.mockClusterRoles.EXPECT().List(v1.ListOptions{LabelSelector: "juju-app==app-name,juju-model==test", IncludeUninitialized: true}).Times(1).
-			Return(&rbacv1.ClusterRoleList{Items: []rbacv1.ClusterRole{*cr}}, nil),
-		s.mockClusterRoles.EXPECT().Update(cr).Times(1).Return(cr, nil),
-		s.mockClusterRoleBindings.EXPECT().List(v1.ListOptions{LabelSelector: "juju-app==app-name,juju-model==test", IncludeUninitialized: true}).Times(1).
-			Return(&rbacv1.ClusterRoleBindingList{Items: []rbacv1.ClusterRoleBinding{*crb}}, nil),
-		s.mockClusterRoleBindings.EXPECT().Delete("read-pods", s.deleteOptions(v1.DeletePropagationForeground, &crbUID)).Times(1).Return(nil),
-		s.mockClusterRoleBindings.EXPECT().Create(crb).Times(1).Return(crb, nil),
-		s.mockSecrets.EXPECT().Update(secretArg).Times(1).Return(nil, s.k8sNotFoundError()),
-		s.mockSecrets.EXPECT().Create(secretArg).Times(1).Return(nil, nil),
+		s.mockRoles.EXPECT().Create(role).Times(1).Return(nil, s.k8sAlreadyExistsError()),
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s_test.go]: To test both code paths, we need to have the mock Create() cal return an AlreadyExists error and then expect a call to Update(). We need to do this wherever ensureFoo() is used.

File: caas/kubernetes/provider/k8s.go [hunk 2]
```diff
@@ -2633,22 +2537,26 @@ func operatorPod(podName, appName, operatorServiceIP, agentPath, operatorImagePa
 
 // operatorConfigMap returns a *core.ConfigMap for the operator pod
 // of the specified application, with the specified configuration.
-func operatorConfigMap(appName, operatorName string, config *caas.OperatorConfig) *core.ConfigMap {
-	configMapName := operatorConfigMapName(operatorName)
+func operatorConfigMap(appName, cmName string, labels map[string]string, config *caas.OperatorConfig) *core.ConfigMap {
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply caas/kubernetes/provider/k8s.go]: it still has knowledge of config data.

File: caas/kubernetes/provider/rbac.go [hunk 3]
```diff
@@ -438,18 +415,23 @@ func (k *kubernetesClient) ensureRoleBinding(rb *rbacv1.RoleBinding) (out *rbacv
 	}
 
 	for _, v := range rbs {
-		if err := k.deleteRoleBinding(v.GetName(), v.GetUID()); err != nil {
-			return nil, cleanups, errors.Trace(err)
+		if v.GetName() == rb.GetName() {
+			if err := k.deleteRoleBinding(v.GetName(), v.GetUID()); err != nil {
+				return nil, cleanups, errors.Trace(err)
+			}
+			logger.Debugf("role binding %q deleted", v.GetName())
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply caas/kubernetes/provider/rbac.go]: ah, right!

File: caas/kubernetes/provider/k8s_test.go (hunk 1 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/k8s_test.go]: yeah, it does return k8sAlreadyExistsError

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
[podSpecV2 - secrets & configmap & sa](https://trello.com/c/Mj3rIu7S/902-podspecv2-secrets-configmap-sa)

**Pull request merged**